### PR TITLE
[DO NOT MERGE] use basic-sender to simplify spec of the then algorithm

### PR DESCRIPTION
**Pull request by ericniebler**
This PR uses the improved _`basic-sender`_ utility to more concisely and precisely specify the `then`, `upon_error`, and `upon_stopped` algorithms.

This is the algorithm template that we can use to re-specify all the other algorithms.

### CODE REVIEW

File: execution.bs [hunk 1]
```diff
@@ -6099,64 +6099,48 @@ template&lt;class Domain, class Tag, sender Sender, class... Args>
...
-                args...)</code> is ill-formed.
-
-            2. <code><i>completion-function</i>(r, args...)</code> is
-                expression-equivalent to <code><i>completion-function</i>(
-                out_r2, args...)</code> when <code><i>completion-function</i></code>
-                is different from <code><i>set-cpo</i></code>.
-
-        2. Let `inner_op` be a non-`const` lvalue referring to the operation
-            state in `o`. `start(o)` is expression-equivalent to `start(inner_op)`.
+4. For `then`, `upon_error`, and `upon_stopped`, let <code><i>set-cpo</i></code>
```
Comment (JeffGarland) [inline execution.bs:6115]: I like this much better than before

File: execution.bs [hunk 2]
```diff
@@ -6099,64 +6099,48 @@ template&lt;class Domain, class Tag, sender Sender, class... Args>
...
+        struct <i>impls-for</i>&lt;<i>then-cpo</i>> : <i>default-impls</i> {
+          static constexpr auto complete =
+            []&lt;class Tag, class... Args>(auto, auto& fn, auto& rcvr, Tag, Args&&... args) noexcept -> void {
+              if constexpr (same_as&lt;Tag, <i>set-cpo</i>>) {
+                <i>TRY-SET-VALUE</i>(std::move(rcvr),
+                              invoke(std::move(fn), std::forward&lt;Args>(args)...));
+              } else {
+                Tag()(std::move(rcvr), std::forward&lt;Args>(args)...);
+              }
+            }
```
Comment (ericniebler) [inline execution.bs:6132]:
```suggestion
            };
```

File: execution.bs (hunk 1 above)
Comment (ericniebler) [reply execution.bs:6115]: i do too. now that i've done one, i think we should create issues to port the other algorithms and assign them to willing volunteers.

File: execution.bs [hunk 3]
```diff
@@ -6099,64 +6099,48 @@ template&lt;class Domain, class Tag, sender Sender, class... Args>
...
+    exposition-only class template <code><i>impls-for</i></code>
+    ([exec.snd.general]) is specialized for each of `then_t`, `upon_error_t`,
+    and `upon_stopped_t` as follows:
+
+        <pre highlight="c++">
+        template&lt;>
+        struct <i>impls-for</i>&lt;tag_t&lt;<i>then-cpo</i>>> : <i>default-impls</i> {
+          static constexpr auto complete =
+            []&lt;class Tag, class... Args>(auto, auto& fn, auto& rcvr, Tag, Args&&... args) noexcept -> void {
+              if constexpr (same_as&lt;Tag, <i>set-cpo</i>>) {
```
Comment (ericniebler) [inline execution.bs:6126]:
```suggestion
              if constexpr (same_as&lt;Tag, tag_t&lt;<i>set-cpo</i>>>) {
```

File: execution.bs [hunk 4]
```diff
@@ -6099,64 +6099,48 @@ template&lt;class Domain, class Tag, sender Sender, class... Args>
...
+              }
+            };
+        };
+        </pre>
 
 5. The expression <code><i>then-cpo</i>(s, f)</code> has undefined behavior
     unless it returns a sender `out_s` that:
+
     1. Invokes `f` or a copy of such with the value, error, or stopped result
         datums of `s` (for `then`, `upon_error`, and `upon_stopped`
```
Comment (mhoemmen) [inline execution.bs:6140]:
```suggestion
        data of `s` (for `then`, `upon_error`, and `upon_stopped`
```

Comment (mhoemmen) [reply execution.bs:6140]: (withdrawn -- thank you for explaining!)

File: execution.bs [hunk 5]
```diff
@@ -6099,64 +6099,48 @@ template&lt;class Domain, class Tag, sender Sender, class... Args>
...
+    be `set_value_t`, `set_error_t`, and `set_stopped_t` respectively. The
+    exposition-only class template <code><i>impls-for</i></code>
+    ([exec.snd.general]) is specialized for each of `then_t`, `upon_error_t`,
+    and `upon_stopped_t` as follows:
+
+        <pre highlight="c++">
+        template&lt;>
+        struct <i>impls-for</i>&lt;tag_t&lt;<i>then-cpo</i>>> : <i>default-impls</i> {
+          static constexpr auto complete =
+            []&lt;class Tag, class... Args>(auto, auto& fn, auto& rcvr, Tag, Args&&... args) noexcept -> void {
```
Comment (mhoemmen) [inline execution.bs:6125]: Would you consider adding a comment to make clear that the first argument matches _`basic-receiver`_'s `Index` template parameter?

```suggestion
            []&lt;class Tag, class... Args>(auto /* index */, auto& fn, auto& rcvr, Tag, Args&&... args) noexcept -> void {
```

**Pull request closed without merging**